nucypher/utilities/porter/porter.py

```diff
@@ -0,0 +1,138 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+from constant_sorrow.constants import NO_CONTROL_PROTOCOL
+from flask import request, Response
+
+from nucypher.blockchain.eth.registry import BaseContractRegistry, InMemoryContractRegistry
+from nucypher.control.controllers import WebController, JSONRPCController
+from nucypher.network.nodes import Learner
+from nucypher.utilities.logging import Logger
+from nucypher.utilities.porter.control.controllers import PorterCLIController
+from nucypher.utilities.porter.control.interfaces import PorterInterface
+
+
+class Porter(Learner):
+
+    BANNER = r"""
+
+ ______
+(_____ \           _
+ _____) )__   ____| |_  ____  ____
+|  ____/ _ \ / ___)  _)/ _  )/ ___)
+| |   | |_| | |   | |_( (/ /| |
+|_|    \___/|_|    \___)____)_|
+
+the conduit between web apps and the nucypher network
+"""
+
+    APP_NAME = "Porter"
+
+    _SHORT_LEARNING_DELAY = 2
+    _LONG_LEARNING_DELAY = 30
+    _ROUNDS_WITHOUT_NODES_AFTER_WHICH_TO_SLOW_DOWN = 25
+
+    DEFAULT_PORTER_HTTP_PORT = 9155
+
+    _interface_class = PorterInterface
+
+    def __init__(self,
+                 domain: str = None,
+                 registry: BaseContractRegistry = None,
+                 controller: bool = True,
+                 *args, **kwargs):
+        self.federated_only = False  # TODO Start with non-federated for now
+        self.registry = registry or InMemoryContractRegistry.from_latest_publication(network=domain)
+
+        super().__init__(save_metadata=True, domain=domain, *args, **kwargs)
+
+        self.log = Logger(self.__class__.__name__)
+
+        # Controller Interface
+        self.interface = self._interface_class(porter=self)
+        self.controller = NO_CONTROL_PROTOCOL
+        if controller:
+            # TODO need to understand this better - only made it analogous to what was done for characters
```

Comment:
Is there any scernerio where you'd want to programmatically use a Porter instance *without* the CLI?